src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContextHelpers.cs

```diff
@@ -36,6 +43,14 @@ internal static void VerifyArguments(Diagnostic diagnostic, Func<Diagnostic, boo
             {
                 throw new ArgumentException(string.Format(CodeAnalysisResources.UnsupportedDiagnosticReported, diagnostic.Id), nameof(diagnostic));
             }
+
+            if (string.IsNullOrEmpty(diagnostic.Id) || !diagnostic.Id.All(UnicodeCharacterUtilities.IsIdentifierPartCharacter))
```

Comment:
You can do that by moving `IsIdentifierStartCharacter` into your new helper class. Better yet, move the rest of the implementation including `IsValidIdentifier` into your new class and just use `IsValidIdentifier` directly. It has precisely the semantics I believe you should be using.